File: bluebook.rb
# BlueBook code decryption (Ruby implementation)

def sieve(n)
  # Compute primes using sieve of Eratosthenes
  x = [1] * n
  x[1] = 0
   
  (2...n/2).each do |i|
    j = 2 * i

    while j < n
      x[j] = 0
      j = j + i
    end
  end

  return x
end

def prime(n, x)
  # Find nth prime
  i = 1
  j = 1
  
  while j <= n
      if x[i] == 1
          j = j + 1
      end

      i = i + 1
  end
  
  return i - 1
end

# Compute BlueBook unlock code
x = sieve(10000)

code = [1206, 301, 384, 5]
key = [1, 1, 2, 2]

$stdout.write [73, 83, 66, 78, 32, 61, 32].map! { |i| i = i.chr }.join("")

(0...4).each do |i|
  $stdout.write(prime(code[i], x) - key[i])
end

puts

File: bluebook.go
// BlueBook code decryption (Go implementation)

package main

import (
    "fmt"
    "strings"
)

// Compute primes using sieve of Eratosthenes
func sieve(n int) []int {
    x := make([]int, n)
    for i, _ := range x {
        x[i] = 1
    }
    x[1] = 0

    for i := 2; i < n/2; i++ {
        j := 2 * i

        for j < n {
            x[j] = 0
            j = j + i
        }
    }

    return x
}

// Find nth prime
func prime(n int, x []int) int {
    i := 1
    j := 1

    for j <= n {
        if x[i] == 1 {
            j++
        }
        i++
    }

    return i - 1
}

// Compute BlueBook unlock code
func main() {
    x := sieve(10000)

    code := []int {1206, 301, 384, 5}
    key := []int {1, 1, 2, 2}

    c := []int {73, 83, 66, 78, 32, 61, 32}
    d := [7]string {}
    for i, char := range c { d[i] = string(char) }
    fmt.Print(strings.Join(d[:], ""))

    for i := 0; i < 4; i++ {
        fmt.Print(prime(code[i], x) - key[i])
    }

    fmt.Println()
}
